<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Data Preparation and Sample Codes for Analysis</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: #990073
   }

   pre .number {
     color: #099;
   }

   pre .comment {
     color: #998;
     font-style: italic
   }

   pre .keyword {
     color: #900;
     font-weight: bold
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: #d14;
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>



<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<!--
%\VignetteEngine{knitr::knitr}
%\VignetteIndexEntry{Sample Code to Use Neoantimon}
-->

<h2>Data Preparation and Sample Codes for Analysis</h2>

<pre><code class="r">#install.packages(&#39;devtools&#39;);
library(devtools);
install_github(&#39;hase62/Neoantimon&#39;);
</code></pre>

<pre><code>## Downloading GitHub repo hase62/Neoantimon@master
## from URL https://api.github.com/repos/hase62/Neoantimon/zipball/master
</code></pre>

<pre><code>## Installing Neoantimon
</code></pre>

<pre><code>## &#39;/Library/Frameworks/R.framework/Resources/bin/R&#39; --no-site-file  \
##   --no-environ --no-save --no-restore --quiet CMD INSTALL  \
##   &#39;/private/var/folders/pc/qs2hrd_j1gv0h5x110_0sn_00000gn/T/RtmpfCStze/devtools54005d462427/hase62-Neoantimon-c3a1e4f&#39;  \
##   --library=&#39;/Library/Frameworks/R.framework/Versions/3.4/Resources/library&#39;  \
##   --install-tests
</code></pre>

<pre><code>## 
</code></pre>

<pre><code>## Reloading installed Neoantimon
</code></pre>

<pre><code class="r">library(Neoantimon);
</code></pre>

<pre><code class="r">data(&quot;sample_vcf&quot;)
head(sample_vcf, row.names = FALSE)
</code></pre>

<pre><code>##   Chr     Start       End Ref Alt Func.refGene Gene.refGene
## 1   1  47399872  47399872   A   G       exonic      CYP4A11
## 2   1 116941338 116941338   T   C       exonic       ATP1A1
## 3   4  24556416  24556416   T   C       exonic        DHX15
## 4   4  70156404  70156404   -   T       exonic      UGT2B28
## 5   6  75899298  75899298   T   -       exonic      COL12A1
## 6   9  89561162  89561162   C   T       exonic         GAS1
##   GeneDetail.refGene ExonicFunc.refGene
## 1      nonsynonymous                SNV
## 2         synonymous                SNV
## 3      nonsynonymous                SNV
## 4         frameshift          insertion
## 5         frameshift           deletion
## 6      nonsynonymous                SNV
##                                                                                                                   AAChange.refGene
## 1                                                                                         CYP4A11:NM_000778:exon8:c.T1064C:p.L355P
## 2 ATP1A1:NM_000701:exon16:c.T2220C:p.D740D,ATP1A1:NM_001160233:exon16:c.T2220C:p.D740D,ATP1A1:NM_001160234:exon16:c.T2127C:p.D709D
## 3                                                                                           DHX15:NM_001358:exon5:c.A1012G:p.T338A
## 4                                                                                      UGT2B28:NM_053039:exon5:c.1186dupT:p.L395fs
## 5                                                                                       COL12A1:NM_004370:exon6:c.628delA:p.I210fs
## 6                                                                                             GAS1:NM_002048:exon1:c.G533A:p.R178H
##   cytoBand depth_tumor variantNum_tumor depth_normal
## 1     1p33          64               28           49
## 2   1p13.1         100               39          111
## 3   4p15.2         143               47          151
## 4   4q13.2          43               15           41
## 5     6q13         122               38           73
## 6  9q21.33          20                5           26
</code></pre>

<pre><code class="r">data(&quot;sample_sv_bnd&quot;)
head(sample_sv_bnd, row.names = FALSE)
</code></pre>

<pre><code>##   Chr     Start       End Ref            Alt Func.refGene Gene.refGene
## 1   1 115005805 115005805   C C]20:34827929]       exonic       TRIM33
## 2   1 204908711 204908711   A [2:172743385[A     intronic        NFASC
## 3   2  25870534  25870534   T  ]2:25965720]G     intronic         DTNB
## 4   2  25965720  25965720   T  T[2:25870536[       exonic        ASXL2
## 5   2 214794791 214794791   C C[2:214798169[       exonic       SPAG16
## 6   2 214798169 214798169   T ]2:214794791]T     intronic       SPAG16
##         mateID
## 1 SVMERGE137_1
## 2  SVMERGE15_1
## 3 SVMERGE116_1
## 4 SVMERGE116_2
## 5   SVMERGE3_1
## 6   SVMERGE3_2
</code></pre>

<pre><code class="r">library(Neoantimon)
data(&quot;sample_hla_table_c1&quot;)
head(sample_hla_table_c1, row.names = FALSE)
</code></pre>

<pre><code>##      Name      A1      A2      B1      B2      C1      C2
## 1  sample A*02:01 A*32:01 B*15:17 B*51:01 C*07:01 C*15:02
## 2 sample2 A*02:01 A*32:01 B*15:17 B*51:01 C*07:01 C*15:02
</code></pre>

<pre><code class="r">data(&quot;sample_hla_table_c2&quot;)
head(sample_hla_table_c2, row.names = FALSE)
</code></pre>

<pre><code>##      Name      DPA11      DPA12      DPB11      DPB12      DQA11
## 1  sample DPA1*01:03 DPA1*02:01 DPB1*02:01 DPB1*09:01 DQA1*01:02
## 2 sample2 DPA1*01:03 DPA1*02:01 DPB1*02:01 DPB1*09:01 DQA1*01:02
##        DQA12      DQB11      DQB12      DRB11      DRB12
## 1 DQA1*05:05 DQB1*03:01 DQB1*06:04 DRB1*11:04 DRB1*13:02
## 2 DQA1*05:05 DQB1*03:01 DQB1*06:04 DRB1*11:04 DRB1*13:02
</code></pre>

<pre><code class="r">data(&quot;sample_rna_exp&quot;)
head(sample_rna_exp, row.names = FALSE)
</code></pre>

<pre><code>##   gene_short_name                            locus expression
## 1             7SK HSCHR6_MHC_MCF:30910595-30910898  0.0000000
## 2             7SK HSCHR6_MHC_QBL:30821624-30821927  0.0000000
## 3             7SK              X:12632121-12632316  0.0000000
## 4            A1BG             19:58856543-58864865  0.1954161
## 5        A1BG-AS1             19:58859116-58866549  5.3048423
## 6            A1CF             10:52559168-52645435  4.7848724
</code></pre>

<pre><code class="r">data(&quot;sample_copynum&quot;)
head(sample_copynum, row.names = FALSE)
</code></pre>

<pre><code>##               Chromosome Position      Log.R segmented.LogR BAF
## SNP_A-8575125          1   564621  0.6071447    -0.09862298   1
## SNP_A-2205441          1   799463  0.1519967    -0.09862298   1
## SNP_A-8325638          1  1017216  0.8146911    -0.09862298   0
## SNP_A-8450345          1  1158277 -1.9594627    -0.09862298   0
## SNP_A-8507155          1  1242215  0.2927962    -0.09862298   0
## SNP_A-8596426          1  1462766 -0.2234090    -0.09862298   1
##               segmented.BAF Copy.number Minor.allele Raw.copy.number
## SNP_A-8575125            NA           2            1       4.3540752
## SNP_A-2205441            NA           2            1       2.1467339
## SNP_A-8325638            NA           3            1       5.8658499
## SNP_A-8450345            NA           1            0      -0.5035897
## SNP_A-8507155            NA           2            1       2.6999875
## SNP_A-8596426            NA           2            1       1.0726687
</code></pre>

<pre><code class="r">data(&quot;sample_result_SNV_CLASS1_ALL&quot;)
head(sample_result_SNV_CLASS1_ALL, row.names = FALSE)
</code></pre>

<pre><code>##           HLA Pos      Gene Evaluated_Mutant_Peptide Mut_IC50 Mut_Rank
## 1 HLA-A*02:01   2 0_CYP4A11            HQERCREEIHSLP  37362.9       55
## 2 HLA-A*02:01   3 0_CYP4A11            QERCREEIHSLPG  45831.8       90
## 3 HLA-A*02:01   4 0_CYP4A11            ERCREEIHSLPGD  45989.7       90
## 4 HLA-A*02:01   5 0_CYP4A11            RCREEIHSLPGDG  46482.0       95
## 5 HLA-A*02:01   6 0_CYP4A11            CREEIHSLPGDGA  42936.1       80
## 6 HLA-A*02:01   7 0_CYP4A11            REEIHSLPGDGAS  44359.4       85
##   Evaluated_Wt_Peptide Wt_IC50 Wt_Rank Chr     NM_ID   Change Ref Alt Prob
## 1        HQERCREEIHSLL 27284.3      32   1 NM_000778 c.T1064C   A   G    0
## 2        QERCREEIHSLLG 44161.6      85   1 NM_000778 c.T1064C   A   G    0
## 3        ERCREEIHSLLGD 44575.4      85   1 NM_000778 c.T1064C   A   G    0
## 4        RCREEIHSLLGDG 44377.6      85   1 NM_000778 c.T1064C   A   G    0
## 5        CREEIHSLLGDGA 41760.9      70   1 NM_000778 c.T1064C   A   G    0
## 6        REEIHSLLGDGAS 43379.7      80   1 NM_000778 c.T1064C   A   G    0
##   Mutation_Prob. Exon_Start Exon_End Mutation_Position Total_Depth
## 1              0   47394845 47407156        1_47399872         113
## 2              0   47394845 47407156        1_47399872         113
## 3              0   47394845 47407156        1_47399872         113
## 4              0   47394845 47407156        1_47399872         113
## 5              0   47394845 47407156        1_47399872         113
## 6              0   47394845 47407156        1_47399872         113
##   Tumor_Depth                      Wt_Peptide
## 1          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 2          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 3          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 4          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 5          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 6          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
##                    Mutant_Peptide Total_RNA Tumor_RNA_Ratio Tumor_RNA
## 1 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 2 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 3 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 4 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 5 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 6 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
##   Tumor_RNA_based_on_DNA MutRatio MutRatio_Min MutRatio_Max
## 1                     NA       NA           NA           NA
## 2                     NA       NA           NA           NA
## 3                     NA       NA           NA           NA
## 4                     NA       NA           NA           NA
## 5                     NA       NA           NA           NA
## 6                     NA       NA           NA           NA
</code></pre>

<pre><code class="r">data(&quot;sample_result_SNV_CLASS2_ALL&quot;)
head(sample_result_SNV_CLASS2_ALL, row.names = FALSE)
</code></pre>

<pre><code>##                     HLA Pos      Gene Evaluated_Mutant_Peptide Mut_IC50
## 1 HLA-DPA10103-DPB10201   1 0_CYP4A11          PKHQERCREEIHSLP  9714.36
## 2 HLA-DPA10103-DPB10201   2 0_CYP4A11          KHQERCREEIHSLPG  9186.08
## 3 HLA-DPA10103-DPB10201   3 0_CYP4A11          HQERCREEIHSLPGD  9621.48
## 4 HLA-DPA10103-DPB10201   4 0_CYP4A11          QERCREEIHSLPGDG 10479.94
## 5 HLA-DPA10103-DPB10201   5 0_CYP4A11          ERCREEIHSLPGDGA 10729.49
## 6 HLA-DPA10103-DPB10201   6 0_CYP4A11          RCREEIHSLPGDGAS 11385.65
##   Mut_Rank Evaluated_Wt_Peptide Wt_IC50 Wt_Rank Chr     NM_ID   Change Ref
## 1       90      PKHQERCREEIHSLL 2610.11      60   1 NM_000778 c.T1064C   A
## 2       90      KHQERCREEIHSLLG 2097.09      55   1 NM_000778 c.T1064C   A
## 3       90      HQERCREEIHSLLGD 1992.73      55   1 NM_000778 c.T1064C   A
## 4       90      QERCREEIHSLLGDG 2154.36      55   1 NM_000778 c.T1064C   A
## 5       90      ERCREEIHSLLGDGA 2193.35      60   1 NM_000778 c.T1064C   A
## 6       95      RCREEIHSLLGDGAS 2362.58      60   1 NM_000778 c.T1064C   A
##   Alt Prob Mutation_Prob. Exon_Start Exon_End Mutation_Position
## 1   G    0              0   47394845 47407156        1_47399872
## 2   G    0              0   47394845 47407156        1_47399872
## 3   G    0              0   47394845 47407156        1_47399872
## 4   G    0              0   47394845 47407156        1_47399872
## 5   G    0              0   47394845 47407156        1_47399872
## 6   G    0              0   47394845 47407156        1_47399872
##   Total_Depth Tumor_Depth                      Wt_Peptide
## 1         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 2         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 3         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 4         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 5         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
## 6         113          64 HPKHQERCREEIHSLLGDGASITWNHLDQMP
##                    Mutant_Peptide Total_RNA Tumor_RNA_Ratio Tumor_RNA
## 1 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 2 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 3 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 4 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 5 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
## 6 HPKHQERCREEIHSLPGDGASITWNHLDQMP        NA              NA        NA
##   Tumor_RNA_based_on_DNA MutRatio MutRatio_Min MutRatio_Max
## 1                     NA       NA           NA           NA
## 2                     NA       NA           NA           NA
## 3                     NA       NA           NA           NA
## 4                     NA       NA           NA           NA
## 5                     NA       NA           NA           NA
## 6                     NA       NA           NA           NA
</code></pre>

<pre><code class="r">data(&quot;sample_result_INDEL_CLASS1_ALL&quot;)
head(sample_result_INDEL_CLASS1_ALL, row.names = FALSE)
</code></pre>

<pre><code>##           HLA Pos      Gene Evaluated_Mutant_Peptide_Core
## 1 HLA-A*02:01   1 0_UGT2B28                    GIPMVGIPLV
## 2 HLA-A*02:01   2 0_UGT2B28                 YHGIPMVGIPLVL
## 3 HLA-A*02:01   3 0_UGT2B28                  HGIPMVGIPLVL
## 4 HLA-A*02:01   4 0_UGT2B28                    GIPMVGIPLV
## 5 HLA-A*02:01   5 0_UGT2B28                    IPMVGIPLVL
## 6 HLA-A*02:01   2 0_UGT2B28                    GIPMVGIPLV
##   Evaluated_Mutant_Peptide Mut_IC50 Mut_Rank Chr     NM_ID         Change
## 1            IYHGIPMVGIPLV   4359.5      8.0   4 NM_053039 Out_c.1186dupT
## 2            YHGIPMVGIPLVL    482.1      3.0   4 NM_053039 Out_c.1186dupT
## 3            HGIPMVGIPLVLG  10645.4     14.0   4 NM_053039 Out_c.1186dupT
## 4            GIPMVGIPLVLGS  10291.2     14.0   4 NM_053039 Out_c.1186dupT
## 5            IPMVGIPLVLGST  13857.7     17.0   4 NM_053039 Out_c.1186dupT
## 6             YHGIPMVGIPLV   3641.2      7.5   4 NM_053039 Out_c.1186dupT
##   Ref Alt Prob Mutation_Prob. Exon_Start Exon_End Mutation_Position
## 1   -   T    0              0   70146216 70160768        4_70156404
## 2   -   T    0              0   70146216 70160768        4_70156404
## 3   -   T    0              0   70146216 70160768        4_70156404
## 4   -   T    0              0   70146216 70160768        4_70156404
## 5   -   T    0              0   70146216 70160768        4_70156404
## 6   -   T    0              0   70146216 70160768        4_70156404
##   Total_Depth Tumor_Depth                     Wt_Peptide
## 1          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 2          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 3          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 4          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 5          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 6          84          43 EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
##         Mutant_Peptide Total_RNA Tumor_RNA_Ratio Tumor_RNA
## 1 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 2 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 3 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 4 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 5 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 6 EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
##   Tumor_RNA_based_on_DNA MutRatio MutRatio_Min MutRatio_Max
## 1                     NA       NA           NA           NA
## 2                     NA       NA           NA           NA
## 3                     NA       NA           NA           NA
## 4                     NA       NA           NA           NA
## 5                     NA       NA           NA           NA
## 6                     NA       NA           NA           NA
</code></pre>

<pre><code class="r">data(&quot;sample_result_INDEL_CLASS2_ALL&quot;)
head(sample_result_INDEL_CLASS2_ALL)
</code></pre>

<pre><code>##                     HLA Pos      Gene Evaluated_Mutant_Peptide_Core
## 1 HLA-DPA10103-DPB10201   0 0_UGT2B28                     YHGIPMVGI
## 2 HLA-DPA10103-DPB10201   1 0_UGT2B28                     YHGIPMVGI
## 3 HLA-DPA10103-DPB10201   2 0_UGT2B28                     MVGIPLVLG
## 4 HLA-DPA10103-DPB10201   3 0_UGT2B28                     MVGIPLVLG
## 5 HLA-DPA10103-DPB10201   4 0_UGT2B28                     MVGIPLVLG
## 6 HLA-DPA10103-DPB10201   0 1_COL12A1                     YQRDELLAA
##   Evaluated_Mutant_Peptide Mut_IC50 Mut_Rank Chr     NM_ID         Change
## 1          EAIYHGIPMVGIPLV  1028.70     40.0   4 NM_053039 Out_c.1186dupT
## 2          AIYHGIPMVGIPLVL  1018.07     40.0   4 NM_053039 Out_c.1186dupT
## 3          IYHGIPMVGIPLVLG  1368.82     46.0   4 NM_053039 Out_c.1186dupT
## 4          YHGIPMVGIPLVLGS  1613.59     49.0   4 NM_053039 Out_c.1186dupT
## 5          HGIPMVGIPLVLGST  2001.80     55.0   4 NM_053039 Out_c.1186dupT
## 6          QYYQRDELLAAIKKF    93.99      8.5   6 NM_004370  Out_c.628delA
##   Ref Alt Prob Mutation_Prob. Exon_Start Exon_End Mutation_Position
## 1   -   T    0              0   70146216 70160768        4_70156404
## 2   -   T    0              0   70146216 70160768        4_70156404
## 3   -   T    0              0   70146216 70160768        4_70156404
## 4   -   T    0              0   70146216 70160768        4_70156404
## 5   -   T    0              0   70146216 70160768        4_70156404
## 6   T   -    0              0   75794041 75915623        6_75899298
##   Total_Depth Tumor_Depth                        Wt_Peptide
## 1          84          43    EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 2          84          43    EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 3          84          43    EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 4          84          43    EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 5          84          43    EAIYHGIPMVGIPLFWDQPDNIAHMKAKGA
## 6         195         122 QYYQRDELLAAIKKIPYKGGNTMTGDAIDYLVK
##            Mutant_Peptide Total_RNA Tumor_RNA_Ratio Tumor_RNA
## 1    EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 2    EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 3    EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 4    EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 5    EAIYHGIPMVGIPLVLGSTX        NA              NA        NA
## 6 QYYQRDELLAAIKKFHIKVATQX        NA              NA        NA
##   Tumor_RNA_based_on_DNA MutRatio MutRatio_Min MutRatio_Max
## 1                     NA       NA           NA           NA
## 2                     NA       NA           NA           NA
## 3                     NA       NA           NA           NA
## 4                     NA       NA           NA           NA
## 5                     NA       NA           NA           NA
## 6                     NA       NA           NA           NA
</code></pre>

<pre><code class="r">TestAnalysis
</code></pre>

<pre><code>## function () 
## {
##     print(&quot;Install NetMHCpan and NetMHCIIpan.&quot;)
##     InstallSampleFiles(url = &quot;https://github.com/hase62/Neoantimon/raw/master/lib/data.zip&quot;, 
##         export_dir = &quot;lib&quot;)
##     InstallSamtools(url = &quot;https://github.com/hase62/Neoantimon/raw/master/lib/samtools-0.1.19.tar.bz2&quot;, 
##         export_dir = &quot;lib&quot;)
##     InstallRefFlat(url = &quot;http://hgdownload.soe.ucsc.edu/goldenPath/hg19/database/refFlat.txt.gz&quot;, 
##         export_dir = &quot;lib&quot;)
##     InstallRefMrnaFile(url = &quot;http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/refMrna.fa.gz&quot;, 
##         export_dir = &quot;lib&quot;)
##     MainSNVClass1(input_file = &quot;lib/data/sample_vcf.txt&quot;, file_name_in_hla_table = &quot;sample&quot;, 
##         hla_file = &quot;lib/data/sample_hla_table_c1.txt&quot;, refflat_file = &quot;lib/refFlat.txt&quot;, 
##         refmrna_file = &quot;lib/refMrna.fa&quot;, rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, 
##         netMHCpan_dir = &quot;lib/netMHCpan-3.0/netMHCpan&quot;, nm_id_column = 10, 
##         depth_tumor_column = 12, depth_normal_column = 14)
##     MainSNVClass2(input_file = &quot;lib/data/sample_vcf.txt&quot;, file_name_in_hla_table = &quot;sample&quot;, 
##         hla_file = &quot;lib/data/sample_hla_table_c2.txt&quot;, refflat_file = &quot;lib/refFlat.txt&quot;, 
##         refmrna_file = &quot;lib/refMrna.fa&quot;, rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, 
##         netMHCIIpan_dir = &quot;lib/netMHCIIpan-3.1/netMHCIIpan&quot;, 
##         nm_id_column = 10, depth_tumor_column = 12, depth_normal_column = 14)
##     MainINDELClass1(input_file = &quot;lib/data/sample_vcf.txt&quot;, file_name_in_hla_table = &quot;sample&quot;, 
##         hla_file = &quot;lib/data/sample_hla_table_c1.txt&quot;, refflat_file = &quot;lib/refFlat.txt&quot;, 
##         refmrna_file = &quot;lib/refMrna.fa&quot;, rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, 
##         netMHCpan_dir = &quot;lib/netMHCpan-3.0/netMHCpan&quot;, nm_id_column = 10, 
##         depth_tumor_column = 12, depth_normal_column = 14)
##     MainINDELClass2(input_file = &quot;lib/data/sample_vcf.txt&quot;, file_name_in_hla_table = &quot;sample&quot;, 
##         hla_file = &quot;lib/data/sample_hla_table_c2.txt&quot;, refflat_file = &quot;lib/refFlat.txt&quot;, 
##         refmrna_file = &quot;lib/refMrna.fa&quot;, rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, 
##         netMHCIIpan_dir = &quot;lib/netMHCIIpan-3.1/netMHCIIpan&quot;, 
##         nm_id_column = 10, depth_tumor_column = 12, depth_normal_column = 14)
##     MainMergeSNVClass1(input_dir = &quot;result.sample.NO_job_id_SNV&quot;, 
##         file_prefix = &quot;NO_job_id_SNV&quot;, annotation_file = &quot;lib/data/sample_vcf.txt.NO_job_id_SNV.peptide.txt&quot;)
##     MainMergeSNVClass2(input_dir = &quot;result.sample.NO_job_id_SNV&quot;, 
##         file_prefix = &quot;NO_job_id_SNV&quot;, annotation_file = &quot;lib/data/sample_vcf.txt.NO_job_id_SNV.peptide.txt&quot;)
##     MainMergeINDELSVClass1(input_dir = &quot;result.sample.NO_job_id_INDEL&quot;, 
##         file_prefix = &quot;NO_job_id_INDEL&quot;, annotation_file = &quot;lib/data/sample_vcf.txt.NO_job_id_INDEL.peptide.txt&quot;)
##     MainMergeINDELSVClass2(input_dir = &quot;result.sample.NO_job_id_INDEL&quot;, 
##         file_prefix = &quot;NO_job_id_INDEL&quot;, annotation_file = &quot;lib/data/sample_vcf.txt.NO_job_id_INDEL.peptide.txt&quot;)
##     MainSVFUSIONClass1(input_file = &quot;lib/data/sample_sv_bnd.txt&quot;, 
##         file_name_in_hla_table = &quot;sample&quot;, hla_file = &quot;lib/data/sample_hla_table_c1.txt&quot;, 
##         refflat_file = &quot;lib/refFlat.txt&quot;, refmrna_file = &quot;lib/refMrna.fa&quot;, 
##         rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, netMHCpan_dir = &quot;lib/netMHCpan-3.0/netMHCpan&quot;, 
##         refdna_file = &quot;lib/GRCh37.fa&quot;, mutation_alt_bnd_column = 5, 
##         gene_symbol_column = 7, mate_id_column = 8)
##     MainSVFUSIONClass2(input_file = &quot;lib/data/sample_sv_bnd.txt&quot;, 
##         file_name_in_hla_table = &quot;sample&quot;, hla_file = &quot;lib/data/sample_hla_table_c2.txt&quot;, 
##         refflat_file = &quot;lib/refFlat.txt&quot;, refmrna_file = &quot;lib/refMrna.fa&quot;, 
##         rnaexp_file = &quot;lib/data/sample_rna_exp.txt&quot;, netMHCIIpan_dir = &quot;lib/netMHCIIpan-3.1/netMHCIIpan&quot;, 
##         refdna_file = &quot;lib/GRCh37.fa&quot;, mutation_alt_bnd_column = 5, 
##         gene_symbol_column = 7, mate_id_column = 8)
##     MainMergeINDELSVClass1(input_dir = &quot;result.sample.NO_job_id_SVFusion&quot;, 
##         file_prefix = &quot;NO_job_id_SVFusion&quot;, annotation_file = &quot;lib/data/sample_sv_bnd.txt.NO_job_id_SVFusion.peptide.txt&quot;)
##     MainMergeINDELSVClass2(input_dir = &quot;result.sample.NO_job_id_SVFusion&quot;, 
##         file_prefix = &quot;NO_job_id_SVFusion&quot;, annotation_file = &quot;lib/data/sample_sv_bnd.txt.NO_job_id_SVFusion.peptide.txt&quot;)
## }
## &lt;environment: namespace:Neoantimon&gt;
</code></pre>

</body>

</html>
